### PR TITLE
MGMT-23736: PublicIP lifecycle enhancements (state wiring, address population, auto-detach)

### PR DESCRIPTION
**Pull request by akshaynadkarni**
## Summary

[MGMT-23736](https://redhat.atlassian.net/browse/MGMT-23736): Complete PublicIP lifecycle enhancements covering three tickets: status.state wiring ([MGMT-24255](https://redhat.atlassian.net/browse/MGMT-24255)), status.address population from MetalLB ([MGMT-24242](https://redhat.atlassian.net/browse/MGMT-24242)), and automatic detachment when ComputeInstances are deleted ([MGMT-23990](https://redhat.atlassian.net/browse/MGMT-23990)).

## Why

The PublicIP CRD defined a State enum (Pending, Allocated, Attaching, Attached, Releasing, Failed) and an address field, but the controllers never populated either. The coarse `status.phase` (Progressing/Ready/Failed/Deleting) cannot distinguish between attach, detach, and allocation flows. Additionally, when a ComputeInstance is deleted, any attached PublicIPs were left in a stale Attached state with no automatic cleanup.

These changes enable the fulfillment service to report precise lifecycle position to consumers, expose the actual IP address allocated by MetalLB, and ensure PublicIPs are automatically released when their ComputeInstance is removed.

### Phase 1: Status.state lifecycle wiring ([MGMT-24255](https://redhat.atlassian.net/browse/MGMT-24255))

**Resource controller** (`publicip_controller.go`):
- Sets `status.state` alongside every existing `status.phase` assignment
- Adds attach/detach detection: Allocated + ComputeInstance set = Attaching, Attached + ComputeInstance cleared = Releasing
- OnSuccess callback checks `spec.computeInstance` to determine Attached vs Allocated
- 8 new unit tests covering all state transitions

**Feedback controller** (`publicip_feedback_controller.go`):
- Replaces `syncPhase` with `syncState` for 1:1 CRD-to-proto state mapping
- Removes all phase-based sync methods (syncPhaseProgressing, syncPhaseFailed, syncPhaseReady, syncPhaseDeleting)
- 8 state mapping test cases (6 states + 2 deletion scenarios)

**Provisioning lifecycle** (`provision_lifecycle.go`):
- Adds `*v1alpha1.PublicIP` case to `GetJobsFromResource`

**BSR version bump** (`buf.gen.yaml`, `internal/api/`):
- Bumps `buf.build/osac-project/private-api` from v0.0.57 to v0.0.59
- Picks up `PUBLIC_IP_STATE_ATTACHING` enum value
- Regenerated gRPC client code: 18 files, ~9,900 lines changed in `internal/api/`. The large diff is from upstream proto additions (role/role_binding types and services) included in the v0.0.59 module, not from this PR's logic changes.

### Phase 2: Status.address population from MetalLB ([MGMT-24242](https://redhat.atlassian.net/browse/MGMT-24242))

**Resource controller** (`publicip_controller.go`):
- Populates `status.address` from the MetalLB LoadBalancer Service's `status.loadBalancer.ingress[0].ip` after successful provisioning
- Clears `status.address` during deprovisioning
- Adds multicluster-aware Service lookup via `multicluster_common.go` helper
- Address population extracted into `maybePopulateAddress` helper

**Supporting changes:**
- `multicluster_common.go`: shared helper for constructing typed clients across hub/remote clusters
- `computeinstance_controller.go`: minor refactor to use shared multicluster helper
- `computeinstance_controller_test.go`: fix flaky envtest reconcile test (race in mock expectations)

### Phase 3: Auto-detach on ComputeInstance deletion ([MGMT-23990](https://redhat.atlassian.net/browse/MGMT-23990))

**Resource controller** (`publicip_controller.go`):
- `handleAutoDetach`: state-based auto-detach handler (Attached clears spec to trigger detach flow, Failed clears stale reference, Attaching waits for in-flight attach, Releasing is no-op, Allocated and default clear spec to prevent finalizer stranding)
- `maybeRemoveCIFinalizer`: removes cross-resource finalizer (`osac.openshift.io/publicip-detach`) from CI only when all PublicIPs have detached (multi-attach safe). Accepts `excludePublicIP` parameter so callers with unpersisted spec changes are not double-counted.
- EventRecorder integration: Normal events for attach/detach, Warning event for auto-detach (generic message for tenant reuse safety). Manual and auto-detach events are deduplicated.

**Constants and RBAC** (`publicip_names.go`, `config/rbac/role.yaml`):
- Controller name, detach finalizer, event reason constants
- MetalLB namespace and service name prefix extracted to constants (prefix updated to `osac-pip-` to match osac-aap convention)
- RBAC markers for CI update, CI finalizers, and events create/patch

**Unit tests** (`publicip_controller_test.go`):
- 8 tests covering all auto-detach scenarios: Attached, Attaching, Releasing, Failed states + multi-attach finalizer lifecycle + CI-not-found edge cases

### Review-driven fixes

The following issues were identified during code review and fixed in follow-up commits:

- **CI detach finalizer never removed**: `priorCIUUID` was captured after `handleAutoDetach` already cleared the spec, so it was always empty. Moved capture to the top of `handleUpdate` before the sync call.
- **Address not populated for one-shot attach**: When a PublicIP is created with `spec.computeInstance` already set, OnSuccess transitions directly to Attached (skipping Allocated). Address is now also populated inside OnSuccess for this path.
- **Finalizer stranded in Allocated/Pending/default states**: These branches did not clear the spec or attempt finalizer removal. Now they clear the spec and call `maybeRemoveCIFinalizer`.
- **Stale-state finalizer scan**: `maybeRemoveCIFinalizer` queried the API server where the spec was not yet persisted, causing false "still referenced" results. Added `excludePublicIP` parameter to skip the current PublicIP during the scan.
- **Duplicate detach event**: Auto-detach emitted both `AutoDetached` and `Detached` events. The `Detached` event is now skipped when the spec change came from auto-detach.

## Testing

```bash
# Build
cd osac-operator && go build ./...

# Full controller test suite (345 tests, 0 failures)
go run github.com/onsi/ginkgo/v2/ginkgo run -v ./internal/controller/

# Auto-detach specific tests (8 tests)
go run github.com/onsi/ginkgo/v2/ginkgo run --focus='auto-detach on ComputeInstance deletion' -v ./internal/controller/

# Lint (0 issues)
make lint
```

## Ticket

[MGMT-23736](https://redhat.atlassian.net/browse/MGMT-23736)

Related tickets:
- [MGMT-24255](https://redhat.atlassian.net/browse/MGMT-24255): Wire status.state transitions
- [MGMT-24242](https://redhat.atlassian.net/browse/MGMT-24242): Populate status.address from MetalLB
- [MGMT-23990](https://redhat.atlassian.net/browse/MGMT-23990): Auto-detach on CI deletion

<br>

<small>Assisted-by: Cursor/Claude</small>

### CODE REVIEW

**Comment by akshaynadkarni** [comment]
@coderabbitai review

**Comment by akshaynadkarni** [comment]
@coderabbitai review

**Comment by akshaynadkarni** [comment]
@coderabbitai review

File: internal/controller/publicip_controller.go [hunk 1]
```diff
@@ -248,14 +262,51 @@ func (r *PublicIPReconciler) handleUpdate(ctx context.Context, publicIP *v1alpha
...
+		publicIP.Status.State = v1alpha1.PublicIPStateAttaching
+	} else if publicIP.Status.State == v1alpha1.PublicIPStateAttached && publicIP.Spec.ComputeInstance == "" {
+		publicIP.Status.Phase = v1alpha1.PublicIPPhaseProgressing
+		publicIP.Status.State = v1alpha1.PublicIPStateReleasing
+		// Emit detach event (Normal for manual detach)
+		if r.Recorder != nil {
+			r.Recorder.Eventf(publicIP, nil, corev1.EventTypeNormal,
+				eventReasonDetached, eventActionReconcile,
+				"PublicIP detached from ComputeInstance")
+		}
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/controller/publicip_controller.go:279]: Fixed in 048759f. The `Detached` event is now skipped when `needsUpdate` is true (meaning auto-detach already emitted `AutoDetached` in this reconcile).

File: internal/controller/publicip_controller.go [hunk 2]
```diff
@@ -314,6 +387,148 @@ func (r *PublicIPReconciler) syncComputeInstanceTargetNamespaceAnnotation(
...
+		// finalizer (added above) from being stranded with no removal path.
+		log.Info("clearing ComputeInstance reference on PublicIP in unexpected state during auto-detach",
+			"state", publicIP.Status.State,
+			"computeInstanceUUID", publicIP.Spec.ComputeInstance)
+		publicIP.Spec.ComputeInstance = ""
+		if err := r.maybeRemoveCIFinalizer(ctx, ci.Labels[osacComputeInstanceIDLabel]); err != nil {
+			return autoDetachResult{}, err
+		}
+		return autoDetachResult{specChanged: true}, nil
+	}
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/controller/publicip_controller.go:488]: Fixed in 048759f. `maybeRemoveCIFinalizer` now takes an `excludePublicIP` parameter. Callers in `handleAutoDetach` (Failed, Allocated, default branches) pass `publicIP.Name` so the method skips the current PublicIP during the API server scan, avoiding the false "still referenced" result from unpersisted spec state. The `OnSuccess` call site passes `""` since the spec is already persisted at that point.

File: internal/controller/publicip_controller.go [hunk 3]
```diff
@@ -451,6 +671,33 @@ func (r *PublicIPReconciler) handleDeprovisioning(ctx context.Context, publicIP
...
+
+	ip := svc.Status.LoadBalancer.Ingress[0].IP
+	if ip == "" {
+		log.Info("LoadBalancer Service ingress IP is empty", "name", serviceName)
+		return ""
+	}
+
+	return ip
+}
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/controller/publicip_controller.go:747]: Fixed in be5685d. Extracted `publicIPServiceNamePrefix` and `defaultMetalLBNamespace` constants into `publicip_names.go`. All hardcoded occurrences in controller and test files replaced. The prefix was also updated from `"osac-publicip-"` to `"osac-pip-"` to match the AAP playbook naming convention in osac-aap.

File: internal/controller/publicip_controller.go [hunk 4]
```diff
@@ -248,11 +257,48 @@ func (r *PublicIPReconciler) handleUpdate(ctx context.Context, publicIP *v1alpha
...
+		if err != nil {
+			log.Error(err, "failed to get target cluster client for address lookup")
+		} else {
+			ipAddress := r.getPublicIPAddress(ctx, targetClient, publicIP.Name)
+			if ipAddress != "" {
+				publicIP.Status.Address = ipAddress
+				log.Info("populated PublicIP address from LoadBalancer Service", "address", ipAddress)
+			}
+		}
 	}
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/controller/publicip_controller.go]: Fixed in 75fce53. Address is now also populated inside the OnSuccess callback when transitioning to Attached, covering the one-shot attach path (Pending -> Attached) that skips the Allocated state.

File: internal/controller/publicip_controller.go [hunk 5]
```diff
@@ -314,6 +382,129 @@ func (r *PublicIPReconciler) syncComputeInstanceTargetNamespaceAnnotation(
...
+		}
+		return autoDetachResult{}, nil
+
+	default:
+		log.Info("unexpected state during auto-detach, skipping",
+			"state", publicIP.Status.State,
+			"computeInstanceUUID", publicIP.Spec.ComputeInstance)
+		return autoDetachResult{}, nil
+	}
+}
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/controller/publicip_controller.go:489]: Fixed in bbaa77d and 048759f. The Allocated case now clears spec.computeInstance and calls maybeRemoveCIFinalizer. The default case (Pending, etc.) now clears the spec and calls maybeRemoveCIFinalizer instead of silently returning. Both pass publicIP.Name via the new excludePublicIP parameter to avoid the stale-state scan issue.

File: internal/controller/publicip_controller.go [hunk 6]
```diff
@@ -344,17 +535,46 @@ func (r *PublicIPReconciler) handleDelete(ctx context.Context, publicIP *v1alpha
...
+					// After detach completes, attempt CI finalizer removal
+					if priorCIUUID != "" {
+						if err := r.maybeRemoveCIFinalizer(ctx, priorCIUUID); err != nil {
+							log.Error(err, "failed to remove CI finalizer after detach",
+								"computeInstanceUUID", priorCIUUID)
+							// Non-fatal: finalizer cleanup will retry on next reconcile
+						}
+					}
+				}
+			},
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/controller/publicip_controller.go:625]: Fixed in 097aa83. `priorCIUUID` is now captured at the top of `handleUpdate` before `syncComputeInstanceTargetNamespaceAnnotation` can clear it via `handleAutoDetach`, and passed as a parameter to `handleProvisioning`. The Failed branch also calls `maybeRemoveCIFinalizer` directly since no AAP detach job runs for that state.

File: internal/controller/multicluster_common.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"context"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	mcmanager "sigs.k8s.io/multicluster-runtime/pkg/manager"
+	mc "sigs.k8s.io/multicluster-runtime/pkg/multicluster"
+)
+
+func getTargetClient(ctx context.Context, mgr mcmanager.Manager, targetCluster mc.ClusterName) (client.Client, error) {
```
Comment (DakCrowder) [inline internal/controller/multicluster_common.go:27]: 🙏

File: internal/controller/publicip_controller.go [hunk 8]
```diff
@@ -343,18 +574,55 @@ func (r *PublicIPReconciler) handleDelete(ctx context.Context, publicIP *v1alpha
...
+						r.Recorder.Eventf(publicIP, nil, corev1.EventTypeNormal,
+							eventReasonAttached, eventActionReconcile,
+							"PublicIP attached to ComputeInstance")
+					}
+					// One-shot attach (Pending -> Attached) skips the Allocated state,
+					// so the address population guard in handleUpdate never fires.
+					// Populate address here to cover that path.
+					if publicIP.Status.Address == "" {
+						if targetClient, err := getTargetClient(ctx, r.mgr, r.targetCluster); err == nil {
+							if ip := r.getPublicIPAddress(ctx, targetClient, publicIP.Name); ip != "" {
```
Comment (DakCrowder) [inline internal/controller/publicip_controller.go:609]: The other caller of `getPublicIPAddress` (`maybePopulateAddress` method) indicates in a comment "One-shot attach (Pending -> Attached) skips Allocated, so that path populates the address inside OnSuccess instead." - which makes total sense.

However, by the time we get here in the OnSuccess and call `getPublicIPAddress`, has the AAP attach job run?  `getPublicIPAddress` has the hardcoded metallb parking namespace `metallb-system` in its lookup and I'm wondering if instead we will need to query the associated VM's namespace to grab the LoadBalancer service?

Comment (akshaynadkarni) [reply internal/controller/publicip_controller.go:609]: Good catch. Today only the `osac-create-public-ip` template runs through the provisioning lifecycle. Neither the AAP nor EDA provider has attach/detach methods wired. The `ProvisioningProvider` interface only defines `TriggerProvision` (create) and `TriggerDeprovision` (delete), so when OnSuccess fires with `spec.computeInstance` set, it's the create job that completed, not an attach job. The Service is still in `metallb-system` at that point, so the lookup happens to work, but the PublicIP is marked Attached without the attach playbook having run.

The `maybePopulateAddress` path (Allocated state) is correct: the create job just finished, the Service is in `metallb-system`, and we read the IP from there.

When attach/detach AAP jobs are wired, the address lookup will need to handle the VM namespace for attached IPs. Filed [OSAC-500](https://redhat.atlassian.net/browse/OSAC-500) to track wiring `osac-attach-public-ip` / `osac-detach-public-ip` into both providers, including updating the address lookup namespace.

**Review by DakCrowder — APPROVED** [review]
